### PR TITLE
Aut-13906 AUT-13945 roll deallocation

### PR DESCRIPTION
**Pull request by sitay93**

### CODE REVIEW

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 1]
```diff
@@ -853,7 +856,38 @@ public void setIsSingleTaskNodesEnabled(Boolean isSingleTaskNodesEnabled) {
     //region Abstract Methods
     abstract List<SpotinstSlave> scaleUp(ProvisionRequest request);
 
-    public abstract Boolean detachInstance(String instanceId);
+    public Boolean removeInstance(String instanceId) {
+        boolean retVal;
+
+        if (isStatefulGroup()) {
+            BaseStatefulInstance statefulInstance   = getStatefulInstance(instanceId);
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: checj if null and write error to log that the group is stateful but the instance id {} doesnt found

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 2]
```diff
@@ -853,22 +857,62 @@ public void setIsSingleTaskNodesEnabled(Boolean isSingleTaskNodesEnabled) {
...
+
+            if(statefulInstance != null && StringUtils.isNotEmpty(statefulInstance.getId())) {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: remove  && StringUtils.isNotEmpty(statefulInstance.getId())

**Review by zivmessing — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 3]
```diff
@@ -853,22 +857,62 @@ public void setIsSingleTaskNodesEnabled(Boolean isSingleTaskNodesEnabled) {
...
+            }
+        }
+        else {
+            retVal = detachInstance(instanceId);
+        }
+
+        return retVal;
+    }
+
+    protected Boolean checkIsStateful(){
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: and rename to checkIsStatefulGroup

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 4]
```diff
@@ -853,22 +857,62 @@ public void setIsSingleTaskNodesEnabled(Boolean isSingleTaskNodesEnabled) {
...
     public abstract String getCloudUrl();
 
-    public void syncGroupInstances() {
+    public void syncGroup() {
         boolean isCloudReadyForGroupCommunication = isCloudReadyForGroupCommunication();
 
         if (isCloudReadyForGroupCommunication) {
-            internalSyncGroupInstances();
+            Boolean isStateful = checkIsStateful();
+
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:910]: bool isSucceeded = isStateful!=null or use blResponse (better)

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 5]
```diff
@@ -853,22 +857,62 @@ public void setIsSingleTaskNodesEnabled(Boolean isSingleTaskNodesEnabled) {
...
+        return retVal;
+    }
+
+    protected Boolean checkIsStateful(){
+        return false;
+    }
+
+    public Boolean isStatefulGroup(){
+        return BooleanUtils.isTrue(isStateful);
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: put it boolean instead if boolean

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 6]
```diff
@@ -50,6 +52,7 @@ public abstract class BaseSpotinstCloud extends Cloud {
     private   EnvironmentVariablesNodeProperty  environmentVariables;
     private   ToolLocationNodeProperty          toolLocations;
     private   Boolean                           shouldUseWebsocket;
+    private   Boolean                           isStateful;
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: change to boolean and initialize as false

File: src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java [hunk 7]
```diff
@@ -214,6 +254,55 @@ protected int getOverriddenNumberOfExecutors(String instanceType) {
...
+        ApiResponse<List<AwsStatefulInstance>> statefulInstancesResponse =
+                awsGroupRepo.getStatefulInstances(groupId, this.accountId);
+
+        if (statefulInstancesResponse.isRequestSucceed()) {
+            statefulInstances = statefulInstancesResponse.getValue();
+        }
+        else {
+            LOGGER.error(String.format("Failed to get group %s stateful instances. Errors: %s", groupId,
+                                       statefulInstancesResponse.getErrors()));
+            statefulInstances = new LinkedList<>();
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java]: remove that line

File: src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java [hunk 8]
```diff
@@ -214,6 +254,55 @@ protected int getOverriddenNumberOfExecutors(String instanceType) {
...
+        }
+
+        Map<String, AwsStatefulInstance> ssiByInstanceId = new HashMap<>();
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java]: put that 3 lines in the if statement

File: src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java [hunk 9]
```diff
@@ -267,6 +356,27 @@ private void addNewSlaveInstances(List<AwsGroupInstance> elastigroupInstances) {
...
+        boolean retVal;
+        boolean isSlaveExist = isSlaveExistForInstance(instance);
+
+        if (isSlaveExist) {
+            retVal = false;
+        }
+        else {
+            if (isStatefulGroup()) {
+                AwsStatefulInstance statefulInstance = getStatefulInstance(instance.getInstanceId());
+                retVal = statefulInstance != null && StringUtils.isNotEmpty(statefulInstance.getId()) &&
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java]: remove StringUtils.isNotEmpty(statefulInstance.getId()

**Review by zivmessing — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/spotinst/common/stateful/StatefulInstance.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+/**
+ * Created by ItayShklar on 07/08/2023.
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class StatefulInstance {
```
Comment (sitay93) [inline src/main/java/hudson/plugins/spotinst/common/stateful/StatefulInstance.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 2 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 3 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 4 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:910]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 5 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 6 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java (hunk 7 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java (hunk 8 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java (hunk 9 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java [hunk 11]
```diff
@@ -214,6 +255,56 @@ protected int getOverriddenNumberOfExecutors(String instanceType) {
...
+
+    private void syncGroupStatefulInstances() {
+        IAwsGroupRepo awsGroupRepo = RepoManager.getInstance().getAwsGroupRepo();
+        ApiResponse<List<AwsStatefulInstance>> statefulInstancesResponse =
+                awsGroupRepo.getStatefulInstances(groupId, this.accountId);
+
+        if (statefulInstancesResponse.isRequestSucceed()) {
+            List<AwsStatefulInstance>        statefulInstances = statefulInstancesResponse.getValue();
+            Map<String, AwsStatefulInstance> ssiByInstanceId   = new HashMap<>();
+            statefulInstances.forEach(ssi -> ssiByInstanceId.put(ssi.getInstanceId(), ssi));
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java]: use toMap just run and test

**Review by zivmessing — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java (hunk 11 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/AwsSpotinstCloud.java]: done

**Review by zivmessing — DISMISSED** [review]

**Review by adivos — APPROVED** [review]

**Pull request merged**